packages/code-studio/src/settings/FormattingSectionContent.tsx

```diff
@@ -201,199 +197,135 @@ export class FormattingSectionContent extends PureComponent<
+      )
+    ) {
+      this.commitChanges(update);
+    }
   }
 
   handleDefaultIntegerFormatChange(event: ChangeEvent<HTMLInputElement>): void {
     log.debug('handleDefaultIntegerFormatChange', event.target.value);
-    this.setState(
-      {
-        defaultIntegerFormatOptions: {
-          defaultFormatString: event.target.value,
-        },
+    const update = {
+      defaultIntegerFormatOptions: {
+        defaultFormatString: event.target.value,
       },
-      () => {
-        this.debouncedCommitChanges();
-      }
-    );
+    };
+    this.setState(update);
+    if (
+      isValidFormat(
+        dh,
+        TableUtils.dataType.INT,
+        IntegerColumnFormatter.makeCustomFormat(event.target.value)
+      )
+    ) {
+      this.commitChanges(update);
+    }
   }
 
   handleShowTimeZoneChange(): void {
-    this.setState(
-      state => ({
-        showTimeZone: !state.showTimeZone,
-      }),
-      () => {
-        this.debouncedCommitChanges();
-      }
-    );
+    const { showTimeZone } = this.state;
+    this.setState({ showTimeZone: !showTimeZone });
+    this.commitChanges({ showTimeZone: !showTimeZone });
   }
 
   handleShowTSeparatorChange(): void {
-    this.setState(
-      state => ({
-        showTSeparator: !state.showTSeparator,
-      }),
-      () => {
-        this.debouncedCommitChanges();
-      }
-    );
+    const { showTSeparator } = this.state;
+    this.setState({ showTSeparator: !showTSeparator });
+    this.commitChanges({ showTSeparator: !showTSeparator });
   }
 
   handleTimeZoneChange(event: ChangeEvent<HTMLSelectElement>): void {
-    this.setState(
-      {
-        timeZone: event.target.value,
-      },
-      () => {
-        this.debouncedCommitChanges();
-      }
-    );
+    this.setState({
+      timeZone: event.target.value,
+    });
+    this.commitChanges({ timeZone: event.target.value });
   }
 
   handleResetDateTimeFormat(): void {
     const { defaults } = this.props;
     const { defaultDateTimeFormat, showTimeZone, showTSeparator } = defaults;
     log.debug('handleResetDateTimeFormat');
-    this.setState(
-      {
-        defaultDateTimeFormat,
-        showTimeZone,
-        showTSeparator,
-      },
-      () => {
-        this.debouncedCommitChanges();
-      }
-    );
+    this.setState({
+      defaultDateTimeFormat,
+      showTimeZone,
+      showTSeparator,
+    });
+    this.debouncedCommitChanges({
+      defaultDateTimeFormat: undefined,
+      showTimeZone: undefined,
+      showTSeparator: undefined,
+    });
   }
 
   handleResetTimeZone(): void {
     const { defaults } = this.props;
     const { timeZone } = defaults;
     log.debug('handleResetTimeZone');
-    this.setState(
-      {
-        timeZone,
-      },
-      () => {
-        this.debouncedCommitChanges();
-      }
-    );
+    this.setState({
+      timeZone,
+    });
+    this.debouncedCommitChanges({
+      timeZone: undefined,
+    });
```

Comment:
Hm. So previously with `debouncedCommitChanges`, it would just get the current state and then set everything in settings... with this change, we're calling `debouncedCommitChanges` with a parameter now of the setting that has changed, and only sending the save for that. The only problem with that is that it's the same `commitChanges` method; so if you change the timezone and then immediately change decimal format afterwards, since the method is debounced it may drop the timezone update and only send the decimal format update (as the first call to `commitChanges` with the timezone param is dropped due to debounce).

To make sure we don't drop any updates, we basically need to keep a queue of all the updates and then apply them. So something like:

```ts
/** Updates that are pending */
pendingUpdates: Partial<WorkspaceSettings>[];

constructor(...) {
  ...
  this.pendingUpdates = []
}

commitChanges(): void {
    // Reduce the updates array to one partial workspace settings update
    // This will merge all the updates into one object
    // This is needed because the debounced function can be called multiple times
    // before the debounce timeout is reached
    const updates = this.pendingUpdates.reduce(
      (acc, update) => ({
        ...acc,
        ...update,
      }),
      {}
    );
    this.pendingUpdates = [];
    
    const { updateSettings } = this.props;
    updateSettings(updates);
}
```